### DISCUSSION TITLE
The color of solid

### DISCUSSION THREAD

**Discussion by RileyGe** [Macad3D/Macad3D#24] [Q&A] [answered]
Can I set the color of a solid? I only find method to change the color of a layer.

**Comment by dipts** [comment, marked as answer]
Currently, a body only receives visual attributes via its layer. There is a [component for the body](https://github.com/Macad3D/Macad3D/blob/main/Source/Macad.Core/Components/VisualStyle.cs) that overwrites these attributes individually - but this has not been used for some time, probably no longer works in all cases and the UI must be re-implemented.

**Reply by RileyGe** [reply]
This is a bad news. I tried to add a VisualStyle to the body, but nothing happened.
That is to say, I cannot change the color and transparency of a body individually?
In addition, I also want to set one edge of a face to be invisible(see the picture below). Is there no way to achieve this through the current code and UI?

![image](https://user-images.githubusercontent.com/26666371/190540689-ab79e55d-ce38-41c9-84f5-888bff553ed0.png)

**Reply by dipts** [reply]
Yes, currently the colour and transparency can only be set by the layer.

The visualisation has no function to hide edges individually. I don't see the point either, if I want to show edges, I want to show all edges - not just some. If the edges within the face are not desired, the faces probably have to be combined. OCCT also has algorithms for this (UnifySameDomain, if I remember correctly).

**Reply by RileyGe** [reply]
Yes, I think this is a good way to solve my problem. So I tried my best to get the right answer, but still I cannot get the right result.

After I add all faces to the compound, I tried the `ShapeUpgrade_UnifySameDomain` class, I think I already do everything I can, but still cannot get the right result.

![image](https://user-images.githubusercontent.com/26666371/190856914-eb53c4b4-80f1-4796-9e33-6e78aced811c.png)

At the end, I get a solid all the same.

![solid image](https://user-images.githubusercontent.com/26666371/190540689-ab79e55d-ce38-41c9-84f5-888bff553ed0.png)

What should I do? Help please.

Thanks

**Reply by dipts** [reply]
Did you try to play with the tolerances? Could be a precision issue.

**Reply by RileyGe** [reply]
I tried play with the tolerance with the code line 158~160, but it has no affects. I think the opencascade maybe has tolerance settings, I am searching for how to dealing with it.

![pic](https://user-images.githubusercontent.com/26666371/190856914-eb53c4b4-80f1-4796-9e33-6e78aced811c.png)

**Reply by RileyGe** [reply]
> Did you try to play with the tolerances? Could be a precision issue.

It is not only the precision issue. I made a mistake, every time I make a face from `DxfPoint`. So even if the two edges are the same, but the `TopoDS_Vetex` is not the same, as a result, the `TopoDS_Edge` is not same. So when I use `ShapeUpgrade_UnifySameDomain` to merge faces, I failed. 
I changed the import method, let all faces have same `TopoDS_Vertex` and `TopoDS_Edge`, the `ShapeUpgrade_UnifySameDomain` get the right result.

Thanks and Regards,
